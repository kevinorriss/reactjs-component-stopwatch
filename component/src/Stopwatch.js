import React, { Fragment } from 'react'
import { CircularProgressbarWithChildren } from 'react-circular-progressbar';
import Moment from 'react-moment'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStopwatch, faUndo, faPlay, faPause } from '@fortawesome/free-solid-svg-icons'
import 'react-circular-progressbar/dist/styles.css'

class Stopwatch extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            counting: false,
            started: 0,
            elapsed: 0,
            lapStarted: 0,
            lapElapsed: 0,
            laps: [],
            timer: null
        }
    }

    tick() {
        console.log('tick()')
        // const now = new Date().getTime()
        //     // calculate the elapsed time since starting
        //     setState(s => ({
        //         ...s,
        //         elapsed: now - s.started,
        //         lapElapsed: now - s.lapStarted
        //     }))
    }

    handlePlayPause() {
        console.log('handlePlayPause()')
        //     // start/resume the timer (adjusting the start time to account for the pause)
        //     if (!counting) {
        //         const now = new Date().getTime()
        //         setState(s => ({
        //             ...s,
        //             counting: true,
        //             started: now - s.elapsed,
        //             lapStarted: now - s.lapElapsed,
        //             timer: setInterval(tick, 10)
        //         }))
        //     // stop timer
        //     } else {
        //         setState(s => ({
        //             ...s,
        //             counting: false,
        //             timer: clearInterval(s.timer)
        //         }))
        //     }
    }

    handleLap() {
        console.log('handleLap()')
        //     setState(s => ({
        //         ...s,
        //         lapStarted: new Date().getTime(),
        //         lapElapsed: 0,
        //         laps: [{
        //             elapsed: s.lapElapsed,
        //             created: s.elapsed
        //         }, ...s.laps]
        //     }))
    }

    handleReset() {
        console.log('handleReset')
        //     setState(s => ({
        //         ...initialState,
        //         timer: clearInterval(s.timer)
        //     }))
    }

    render() {
        return (
            <Fragment>
            <div className="pl-5 pr-5">
                <CircularProgressbarWithChildren value={this.state.elapsed % 60000} maxValue={60000} className="d-felx align-items-center" styles={{path:{transition: 'none'}}}>
                    <h3 className="mt-4"><Moment format="mm:ss.SS">{this.state.elapsed}</Moment></h3>
                    <p className="mb-0"><Moment format="mm:ss.SS">{this.state.lapElapsed}</Moment></p>
                </CircularProgressbarWithChildren>
            </div>
            <div className="stopwatch-laps flex-grow-1">
                {this.state.laps.map((lap, index) => (
                    <div key={index}>
                        <div className="d-flex">
                            <div className="col-6 d-flex flex-column ml-1">
                                <div className="font-weight-bold" style={{ lineHeight: '1rem'}}>Lap {this.state.laps.length - index}</div>
                                <div style={{ lineHeight: '1rem'}}><small>Elapsed <Moment format="mm:ss.SS">{lap.elapsed}</Moment></small></div>
                            </div>
                            <div className="col-6 d-flex flex-row-reverse align-items-center mr-1">
                                <h4 className="font-weight-bold mb-0"><Moment format="mm:ss.SS">{lap.created}</Moment></h4>
                            </div>
                        </div>
                        <hr className="mt-1 mb-1"/>
                    </div>
                ))}
            </div>
            <div className="d-flex justify-content-around ml-5 mr-5">
                <button className="btn" onClick={this.handleReset}>
                    <FontAwesomeIcon icon={faUndo}/>
                </button>
                <button className="btn btn-light border shadow-sm" onClick={this.handlePlayPause}>
                    <FontAwesomeIcon icon={this.state.counting ? faPause : faPlay} />
                </button>
                <button className="btn" onClick={this.onLap} disabled={!this.state.counting}>
                    <FontAwesomeIcon icon={faStopwatch} />
                </button>
            </div>
        </Fragment>
        )
    }
}

export default Stopwatch