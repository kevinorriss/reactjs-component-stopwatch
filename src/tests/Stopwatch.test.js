import React from 'react'
import { mount } from 'enzyme'
import moment from 'moment'
import renderer from 'react-test-renderer'
import Stopwatch from '../../component/src/Stopwatch'

let wrapper
beforeEach(() => {
    wrapper = mount(<Stopwatch/>)
})

const sleep = (milliseconds) => {
    return new Promise(resolve => setTimeout(resolve, milliseconds));
}

const elapsedToMilliseconds = (elapsedText) => {
    return moment(elapsedText, 'mm:ss.S').diff(moment().startOf('day'), 'milliseconds')
}

const getElapsedText = () => {
    return wrapper.find('.face__time--elapsed').first().text()
}

const getLapElapsedText = () => {
    return wrapper.find('.face__time--lap-elapsed').first().text()
}

const getElapsedMilliseconds = () => {
    return elapsedToMilliseconds(getElapsedText())
}

const getLapElapsedMilliseconds = () => {
    return elapsedToMilliseconds(getLapElapsedText())
}

const clickPlayPause = () => {
    wrapper.find('.controls__button--play').simulate('click')
}

const clickLap = () => {
    wrapper.find('.controls__button--lap').simulate('click')
}

const clickReset = () => {
    wrapper.find('.controls__button--reset').simulate('click')
}

it('should render Stopwatch correctly', () => {
    expect(renderer.create(<Stopwatch/>).toJSON()).toMatchSnapshot()
})

it('should start the timer', async () => {
    // expect a zero start
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)

    // press play
    clickPlayPause()

    // wait half a second
    await sleep(500)

    // get the elapsed time
    const elapsed = getElapsedMilliseconds()

    // expect a non-zero elapsed time
    expect(elapsed).toBeGreaterThan(0)
    expect(getLapElapsedMilliseconds()).toEqual(elapsed)
})

it('should enable lap button on play', async () => {
    // expect the lap button to be disabled
    expect(wrapper.find('.controls__button--lap').first().prop('disabled')).toBeTruthy()
    
    // press play
    clickPlayPause()

    // wait half a second
    await sleep(500)

    // expect the lap button to be disabled
    expect(wrapper.find('.controls__button--lap').first().prop('disabled')).toBeFalsy()
})

it('should disable lap button on pause', async () => {
    // press play
    clickPlayPause()

    // wait half a second
    await sleep(500)

    // press play
    clickPlayPause()

    // wait half a second
    await sleep(500)

    // expect the lap button to be disabled
    expect(wrapper.find('.controls__button--lap').first().prop('disabled')).toBeTruthy()
})

it('should stop the timer', async () => {
    // expect a zero start
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)

    // press play
    clickPlayPause()

    // wait half a second
    await sleep(500)

    // press pause
    clickPlayPause()

    // get the current elapsed time
    const milliseconds = getElapsedMilliseconds()
    const lapMilliseconds = getLapElapsedMilliseconds()

    // wait another half a second
    await sleep(500)

    // elapsed time should have stayed the same
    expect(getElapsedMilliseconds()).toEqual(milliseconds)
    expect(getLapElapsedMilliseconds()).toEqual(lapMilliseconds)
})

it('should resume the timer', async () => {
    // expect a zero start
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)

    // play
    clickPlayPause()
    await sleep(500)

    // pause
    clickPlayPause()
    await sleep(500)

    // get the elapsed time
    const milliseconds = getElapsedMilliseconds()
    const lapMilliseconds = getLapElapsedMilliseconds()

    // play
    clickPlayPause()
    await sleep(500)

    // elapsed time should have increased
    expect(getElapsedMilliseconds()).toBeGreaterThan(milliseconds)
    expect(getLapElapsedMilliseconds()).toBeGreaterThan(lapMilliseconds)
})

it('should add laps', async () => {
    // expect a zero start
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)

    // create two laps
    clickPlayPause()
    await sleep(1000)
    clickLap()
    await sleep(1000)
    clickLap()
    clickPlayPause()

    // expect the total elapsed to be higher than the current lap elapsed
    expect(getElapsedMilliseconds()).toBeGreaterThan(getLapElapsedMilliseconds())

    // expect two lap elements
    const laps = wrapper.find('.stopwatch__laps').first().children()
    expect(laps.length).toEqual(2)

    // expect lap one
    const lapOne = laps.at(1)
    expect(lapOne.find('.entry__elapsed h4').text()).toEqual('Lap 1')
    const lapOneElapsed = elapsedToMilliseconds(lapOne.find('.entry__elapsed small').text())
    expect(lapOneElapsed).toBeGreaterThan(0)
    const lapOneCreated = elapsedToMilliseconds(lapOne.find('.entry__created').text())
    expect(lapOneCreated).toEqual(lapOneElapsed)

    // expect lap two
    const lapTwo = laps.at(0)
    expect(lapTwo.find('.entry__elapsed h4').text()).toEqual('Lap 2')
    const lapTwoElapsed = elapsedToMilliseconds(lapTwo.find('.entry__elapsed small').text())
    expect(lapTwoElapsed).toBeGreaterThan(0)
    expect(elapsedToMilliseconds(lapTwo.find('.entry__created').text())).toBeGreaterThan(lapOneCreated)
})

it('should reset the stopwatch', async () => {
    // expect a zero start
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)

    // create two laps, reset and wait
    clickPlayPause()
    await sleep(100)
    clickLap()
    await sleep(100)
    clickLap()
    clickReset()
    await sleep(500)

    // expect 0 elapsed and no laps
    expect(getElapsedMilliseconds()).toEqual(0)
    expect(getLapElapsedMilliseconds()).toEqual(0)
    expect(wrapper.find('.stopwatch__laps').first().children().length).toEqual(0)
    expect(wrapper.find('.controls__button--lap').first().prop('disabled')).toBeTruthy()
})