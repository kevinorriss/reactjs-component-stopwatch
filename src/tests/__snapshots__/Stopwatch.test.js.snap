// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Stopwatch correctly 1`] = `
<div
  className="stopwatch"
>
  <div
    className="stopwatch__face"
  >
    <div
      data-test-id="CircularProgressbarWithChildren"
    >
      <div
        style={
          Object {
            "height": "100%",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <svg
          className="CircularProgressbar "
          data-test-id="CircularProgressbar"
          viewBox="0 0 100 100"
        >
          <path
            className="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fillOpacity={0}
            strokeWidth={8}
            style={
              Object {
                "strokeDasharray": "289.02652413026095px 289.02652413026095px",
                "strokeDashoffset": "0px",
              }
            }
          />
          <path
            className="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fillOpacity={0}
            strokeWidth={8}
            style={
              Object {
                "strokeDasharray": "289.02652413026095px 289.02652413026095px",
                "strokeDashoffset": "289.02652413026095px",
                "transition": "none",
              }
            }
          />
        </svg>
        <div
          data-test-id="CircularProgressbarWithChildren__children"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "column",
              "height": "100%",
              "justifyContent": "center",
              "marginTop": "-100%",
              "position": "absolute",
              "width": "100%",
            }
          }
        >
          <time
            className="face__time--elapsed"
            dateTime={"1970-01-01T00:00:00.000Z"}
          >
            00:00.0
          </time>
          <time
            className="face__time--lap-elapsed"
            dateTime={"1970-01-01T00:00:00.000Z"}
          >
            00:00.0
          </time>
        </div>
      </div>
    </div>
  </div>
  <div
    className="stopwatch__laps"
  />
  <div
    className="stopwatch__controls"
  >
    <button
      className="controls__button controls__button--reset"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-undo fa-w-16 "
        data-icon="undo"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <button
      className="controls__button controls__button--play"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-play fa-w-14 "
        data-icon="play"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <button
      className="controls__button controls__button--lap"
      disabled={true}
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-stopwatch fa-w-14 "
        data-icon="stopwatch"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M432 304c0 114.9-93.1 208-208 208S16 418.9 16 304c0-104 76.3-190.2 176-205.5V64h-28c-6.6 0-12-5.4-12-12V12c0-6.6 5.4-12 12-12h120c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-28v34.5c37.5 5.8 71.7 21.6 99.7 44.6l27.5-27.5c4.7-4.7 12.3-4.7 17 0l28.3 28.3c4.7 4.7 4.7 12.3 0 17l-29.4 29.4-.6.6C419.7 223.3 432 262.2 432 304zm-176 36V188.5c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12V340c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
</div>
`;
